From: Haifeng Xu <haifeng.xu@shopee.com>
Subject: memcg, oom: unmark under_oom after the oom killer is done
Date: Sat, 23 Sep 2023 08:17:39 +0000

When application in userland receives oom notification from kernel and
reads the oom_control file, it's confusing that under_oom is 0 though the
omm killer hasn't finished.  The reason is that under_oom is cleared
before invoking mem_cgroup_out_of_memory(), so move the action that unmark
under_oom after completing oom handler.  Therefore the value of under_oom
won't mislead users.

Link: https://lkml.kernel.org/r/20230923081739.398912-1-haifeng.xu@shopee.com
Signed-off-by: Haifeng Xu <haifeng.xu@shopee.com>
Acked-by: Roman Gushchin <roman.gushchin@linux.dev>
Cc: Johannes Weiner <hannes@cmpxchg.org>
Cc: Michal Hocko <mhocko@kernel.org>
Cc: Shakeel Butt <shakeelb@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/memcontrol.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/mm/memcontrol.c~memcg-oom-unmark-under_oom-after-the-oom-killer-is-done
+++ a/mm/memcontrol.c
@@ -2027,8 +2027,8 @@ static bool mem_cgroup_oom(struct mem_cg
 	if (locked)
 		mem_cgroup_oom_notify(memcg);
 
-	mem_cgroup_unmark_under_oom(memcg);
 	ret = mem_cgroup_out_of_memory(memcg, mask, order);
+	mem_cgroup_unmark_under_oom(memcg);
 
 	if (locked)
 		mem_cgroup_oom_unlock(memcg);
_
